modules/twitter/src/main/java/org/apache/ignite/stream/twitter/TwitterStreamer.java

```diff
@@ -0,0 +1,280 @@
+ * limitations under the License.
+ */
+
+package org.apache.ignite.stream.twitter;
+
+import com.google.common.base.Splitter;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.twitter.hbc.ClientBuilder;
+import com.twitter.hbc.core.Client;
+import com.twitter.hbc.core.HttpConstants;
+import com.twitter.hbc.core.HttpHosts;
+import com.twitter.hbc.core.endpoint.DefaultStreamingEndpoint;
+import com.twitter.hbc.core.endpoint.SitestreamEndpoint;
+import com.twitter.hbc.core.endpoint.StatusesFilterEndpoint;
+import com.twitter.hbc.core.endpoint.StatusesFirehoseEndpoint;
+import com.twitter.hbc.core.endpoint.StatusesSampleEndpoint;
+import com.twitter.hbc.core.endpoint.StreamingEndpoint;
+import com.twitter.hbc.core.endpoint.UserstreamEndpoint;
+import com.twitter.hbc.core.processor.StringDelimitedProcessor;
+import com.twitter.hbc.httpclient.auth.Authentication;
+import com.twitter.hbc.httpclient.auth.OAuth1;
+import org.apache.ignite.IgniteDataStreamer;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.util.typedef.internal.A;
+import org.apache.ignite.stream.StreamAdapter;
+import twitter4j.Status;
+import twitter4j.TwitterObjectFactory;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Streamer that consumes from a Twitter Streaming API and feeds transformed key-value pairs, by default < tweetId, text>,
+ * into an {@link IgniteDataStreamer} instance.
+ * <p>
+ * This streamer uses https://dev.twitter.com/streaming API and supports Public API, User Streams, Site Streams and Firehose.
+ * <p>
+ * You can also provide a {@link TweetTransformer} to convert the incoming message into cache entries to override
+ * default transformer.
+ * <p>
+ * This Streamer features:
+ *
+ * <ul>
+ *     <li>Supports OAuth1 authentication scheme. <br/> BasicAuth not supported by Streaming API https://dev.twitter.com/streaming/overview/connecting</li>
+ *     <li>Provide all params in apiParams map. https://dev.twitter.com/streaming/overview/request-parameters</li>
+ * </ul>
+ *
+ */
+public class TwitterStreamer<K, V> extends StreamAdapter<String, K, V> {
+
+    /** Logger. */
+    private IgniteLogger log;
+
+    /**
+     * The message transformer that converts an incoming Tweet into cache entries.
+     * If not provided default transformer will be used.
+     */
+    private TweetTransformer<K, V> transformer;
+
+    /** Twitter Streaming API params. See https://dev.twitter.com/streaming/overview/request-parameters */
+    private Map<String, String> apiParams = new HashMap<String, String>();
```

Comment:
As this param is optional, so initializing it with Collections.EMPTY_MAP to avoid null checks